# 30 Pandas Exercises With Real Hospital Data (With Hints & Solutions)<a id='erd'></a> 

![cover](https://i.ibb.co/WWBMJ46/cover1600.png)

To see the proposed solutions, visit my [Medium article here]().

## Importing Pandas:

In [1]:
import pandas as pd

The following optional cell applies some CSS formatting to Pandas dataframes to add more visual contrast between the indices and the data. It takes effect immediately after you run it. For more information, you can read my <u>[Medium article about it here](https://medium.com/analytics-vidhya/5-css-snippets-for-better-looking-pandas-dataframes-ad808e407894)</u>.

In [2]:
%%HTML
<style>.dataframe th{background:#3f577c;font-family:monospace;color:white;border:3px solid white;text-align:left !important;
}.dataframe td{text-align:left;min-width:5em;}.column{column-count: 3;}</style>

If you don't like the CSS style, simply select the code inside of the cell, hit **`CTRL + /`** to turn it into a comment (or add **`#`** at the start of each line), and then run the cell.

---

## The Data - Real Hospital Records: <a href="" id="dataschema"></a>

#### <strong style="color:#85e468;font-size:100%">⦿ </strong>Patients (974 records)

In [3]:
patients = pd.read_csv('data/patients.csv')

#### <strong style="color:#486ce3;font-size:100%">⦿ </strong> Payers (10 records)

In [4]:
payers = pd.read_csv('data/payers.csv')

#### <strong style="color:#e45151;font-size:100%">⦿ </strong> Encounters (27,891 records)

In [5]:
encounters = pd.read_csv('data/encounters.csv')

#### <strong style="color:#ca4fec;font-size:100%">⦿ </strong> Procedures (47,701 records)

In [6]:
procedures = pd.read_csv('data/procedures.csv')

The following schema can help you get a sense of the data we will be working with.

![schema](https://i.ibb.co/GRz4v2d/schema.png)

---

## The Data Dictionary:

The following is a data dictionary explaining what each dataframe and column/variable means:

### <strong style="color:#85e468;font-size:100%">⦿ </strong>Patients (974 records)

In [110]:
patients.head(1)

PATIENT_ID BIRTH_DATE DEATH_DATE FIRST_NAME  \
0  5605b66b-e92d-c16c-1b83-b8bf7040d51f  3/19/1977        NaN     Nikita   

  LAST_NAME MARITAL   RACE    ETHNICITY GENDER                    BIRTHPLACE  \
0    Erdman       M  white  nonhispanic      F  Wakefield  Massachusetts  US   

              PATIENT_ADDRESS PATIENT_CITY  PATIENT_STATE  PATIENT_COUNTY  \
0  510 Little Station Unit 69       Quincy  Massachusetts  Norfolk County   

   PATIENT_ZIP        LAT        LON  
0       2186.0  42.290937 -70.975503

The Patients dataframe contains demographic information about individuals receiving care. Each entry in the dataframe represents a unique patient, identified by a primary key, PATIENT_ID. The dataframe includes details such as the patient's birth date, potential death date, name, marital status, race, ethnicity, gender, birthplace, and address information.

<style>
.table_component {
    overflow: auto;
    width: 100%;
    text-align: left;
}

.table_component table {
    height: 100%;
    width: 100%;
    table-layout: auto;
    border-collapse: collapse;
    border-spacing: 1px;
    text-align: left;
}

.table_component caption {
    caption-side: top;
    text-align: left;
}

.table_component th {
    border: 1px solid #dededf;
    background-color: #eceff1;
    color: #000000;
    padding: 5px;
    text-align: left;
}

.table_component td {
    border: 1px solid #dededf;
    background-color: #ffffff;
    color: #000000;
    padding: 5px;
    text-align: left;
}
</style>

<div class="table_component" role="region" tabindex="0">
<table style="border:solid 5px #85e468;">
    <caption>
        <p></p>
    </caption>
    <thead>
        <tr>
            <th>Column/Attribute</th>
            <th>Description</th>
            <th>Example</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>
                <ul>
                    <li><strong>PATIENT_ID</strong></li>
                </ul>
            </td>
            <td>Primary Key. Unique Identifier of the patient &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp;</td>
            <td>
                <div>f96addf5-81b9-0aab-7855-d208d3d352c5 &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp;</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>BIRTH_DATE</strong></li>
                </ul>
            </td>
            <td>The date the patient was born (YYYY-MM-DD)</td>
            <td>
                <div>12/25/1928</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>DEATH_DATE</strong></li>
                </ul>
            </td>
            <td>The date the patient died (YYYY-MM-DD)</td>
            <td>
                <div>2/23/2014</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>FIRST_NAME</strong></li>
                </ul>
            </td>
            <td>First name of the patient</td>
            <td>
                <div>Edwin</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>LAST_NAME</strong></li>
                </ul>
            </td>
            <td>Last or surname of the patient</td>
            <td>
                <div>Labadie</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>MARITAL</strong></li>
                </ul>
            </td>
            <td>Marital Status. M is married, S is single</td>
            <td>
                <div>M</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>RACE</strong></li>
                </ul>
            </td>
            <td>Description of the patient's primary race</td>
            <td>
                <div>white</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>ETHNICITY</strong></li>
                </ul>
            </td>
            <td>Description of the patient's primary ethnicity</td>
            <td>
                <div>hispanic</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>GENDER</strong></li>
                </ul>
            </td>
            <td>Gender. M is male, F is female</td>
            <td>
                <div>M</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>BIRTHPLACE</strong></li>
                </ul>
            </td>
            <td>Name of the town where the patient was born</td>
            <td>
                <div>Stow Massachusetts US</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>PATIENT_ADDRESS</strong></li>
                </ul>
            </td>
            <td>Patient's street address without commas or newlines</td>
            <td>
                <div>976 Ziemann Gateway</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>PATIENT_CITY</strong></li>
                </ul>
            </td>
            <td>Patient's address city</td>
            <td>
                <div>Boston</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>PATIENT_STATE</strong></li>
                </ul>
            </td>
            <td>Patient's address state</td>
            <td>
                <div>Massachusetts</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>PATIENT_COUNTY</strong></li>
                </ul>
            </td>
            <td>Patient's address county</td>
            <td>
                <div>Suffolk County</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>PATIENT_ZIP</strong></li>
                </ul>
            </td>
            <td>Patient's zip code</td>
            <td>
                <div>2125</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>LAT</strong></li>
                </ul>
            </td>
            <td>Latitude of Patient's address</td>
            <td>
                <div>42.3467714</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>LON</strong></li>
                </ul>
            </td>
            <td>Longitude of Patient's address</td>
            <td>
                <div>-71.05881297</div>
            </td>
        </tr>
    </tbody>
</table>
</div>

### <strong style="color:#486ce3;font-size:100%">⦿ </strong> Payers (10 records) <a href="" id="#payersdict"></a>


In [111]:
payers.head(1)

PAYER_ID     PAYER_NAME       PAYER_ADDRESS  \
0  b3221cfc-24fb-339e-823d-bc4136cbc4ed  Dual Eligible  7500 Security Blvd   

  PAYER_CITY STATE_HEADQUARTERED  PAYER_ZIP     PAYER_PHONE  
0  Baltimore                  MD    21244.0  1-877-267-2323

The Payers dataframe lists entities responsible for covering healthcare costs, such as insurance companies. Each payer is uniquely identified by the PAYER_ID. The dataframe provides the name of the payer, their street address, city, state abbreviation, zip code, and phone number.

<div class="table_component" role="region" tabindex="0">
<table style="border:solid 5px #486ce3;">
    <caption>
        <p></p>
    </caption>
    <thead>
        <tr>
            <th>Column/Attribute</th>
            <th>Description</th>
            <th>Example</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>
                <ul>
                    <li><strong>PAYER_ID</strong></li>
                </ul>
            </td>
            <td>Primary key of the Payer (e.g. Insurance)</td>
            <td>
                <div>b3221cfc-24fb-339e-823d-bc4136cbc4ed</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>PAYER_NAME</strong></li>
                </ul>
            </td>
            <td>Name of the Payer</td>
            <td>
                <div>Dual Eligible</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>PAYER_ADDRESS</strong></li>
                </ul>
            </td>
            <td>Payer's street address without commas or newlines &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp;</td>
            <td>
                <div>7500 Security Blvd &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp;</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>PAYER_CITY</strong></li>
                </ul>
            </td>
            <td> Street address city</td>
            <td>
                <div>Baltimore</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>STATE_HEADQUARTERED</strong></li>
                </ul>
            </td>
            <td>Street address state abbreviation</td>
            <td>
                <div>MD</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>PAYER_ZIP</strong></li>
                </ul>
            </td>
            <td>Street address zip or postal code</td>
            <td>
                <div>21244</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>PAYER_PHONE</strong></li>
                </ul>
            </td>
            <td>Payer's phone number</td>
            <td>
                <div>1-877-267-2323</div>
            </td>
        </tr>
    </tbody>
</table>
</div>

### <strong style="color:#e45151;font-size:100%">⦿ </strong> Encounters (27,891 records)

In [103]:
encounters.head(1)

ENCOUNTER_ID                            PATIENT_ID  \
0  32c84703-2481-49cd-d571-3899d5820253  3de74169-7f67-9304-91d4-757e0f3a14d2   

                               PAYER_ID       ENCOUNTER_START  \
0  b1c428d6-4f07-31e0-90f0-68ffa6ff8c76  2011-01-02T09:26:36Z   

         ENCOUNTER_STOP ENCOUNTER_CLASS  ENCOUNTER_CODE  \
0  2011-01-02T12:58:36Z      ambulatory       185347001   

               ENCOUNTER_DESCRIPTION  ENCOUNTER_BASE_COST  TOTAL_CLAIM_COST  \
0  Encounter for problem (procedure)                85.55           1018.02   

   PAYER_COVERAGE  ENCOUNTER_REASON_CODE ENCOUNTER_REASON_DESCRIPTION  
0             0.0                    NaN                          NaN

The Encounters dataframe records interactions between patients and the hospital providing healthcare. Each encounter is uniquely identified by an ENCOUNTER_ID. The dataframe includes foreign keys linking to the corresponding patient (PATIENT_ID) and payer (PAYER_ID). It documents the start and stop times of the encounter, the class and description of the encounter, associated codes from SNOMED-CT, the base and total costs of the encounter, payer coverage, and any specific reasons for the encounter identified by diagnosis codes and descriptions.

<div class="table_component" role="region" tabindex="0">
<table style="border:solid 5px #e45151;">
    <caption>
        <p></p>
    </caption>
    <thead>
        <tr>
            <th>Column/Attribute</th>
            <th>Description</th>
            <th>Example</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>
                <ul>
                    <li><strong>ENCOUNTER_ID</strong></li>
                </ul>
            </td>
            <td>Primary Key. Unique Identifier of the encounter</td>
            <td>
                <div>5c504a86-1804-ed37-53ce-32f3d3318ad5</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>PATIENT_ID</strong></li>
                </ul>
            </td>
            <td>Foreign key to the Patient</td>
            <td>
                <div>8fe3348d-477f-541e-96df-bbe5b552104f</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>PAYER_ID</strong></li>
                </ul>
            </td>
            <td>Foreign key to the Payer</td>
            <td>
                <div>7c4411ce-02f1-39b5-b9ec-dfbea9ad3c1a</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>ENCOUNTER_START</strong></li>
                </ul>
            </td>
            <td> The date and time the encounter started (iso8601 UTC Date (yyyy-MM-dd'T'HH:mm'Z'))</td>
            <td>
                <div>2011-01-13T15:41:08Z</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>ENCOUNTER_STOP</strong></li>
                </ul>
            </td>
            <td>The date and time the encounter concluded (iso8601 UTC Date (yyyy-MM-dd'T'HH:mm'Z'))</td>
            <td>
                <div>2011-01-13T15:56:08Z</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>ENCOUNTER_CLASS</strong></li>
                </ul>
            </td>
            <td>The class of the encounter, such as ambulatory, emergency, inpatient, wellness, or urgent care</td>
            <td>
                <div>ambulatory</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>ENCOUNTER_CODE</strong></li>
                </ul>
            </td>
            <td>Encounter code from SNOMED-CT</td>
            <td>
                <div>424619006</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>ENCOUNTER_DESCRIPTION</strong></li>
                </ul>
            </td>
            <td>Description of the type of encounter</td>
            <td>
                <div>Prenatal visit</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>ENCOUNTER_BASE_COST</strong></li>
                </ul>
            </td>
            <td>The base cost of the encounter, not including any line item costs related to medications, immunizations, procedures, or other services</td>
            <td>
                <div>142.58</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>TOTAL_CLAIM_COST</strong></li>
                </ul>
            </td>
            <td>The total cost of the encounter, including all line items</td>
            <td>
                <div>13281.55</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>PAYER_COVERAGE</strong></li>
                </ul>
            </td>
            <td>The amount of cost covered by the Payer</td>
            <td>
                <div>12560.47</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>ENCOUNTER_REASON_CODE</strong></li>
                </ul>
            </td>
            <td>Diagnosis code from SNOMED-CT, only if this encounter targeted a specific condition</td>
            <td>
                <div>72892002</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>ENCOUNTER_REASON_DESCRIPTION</strong></li>
                </ul>
            </td>
            <td>Description of the reason code</td>
            <td>
                <div>Normal pregnancy</div>
            </td>
        </tr>
    </tbody>
</table>
</div>

### <strong style="color:#ca4fec;font-size:100%">⦿ </strong> Procedures (47,701 records)

In [104]:
procedures.head(1)

PATIENT_ID                          ENCOUNTER_ID  \
0  3de74169-7f67-9304-91d4-757e0f3a14d2  32c84703-2481-49cd-d571-3899d5820253   

        PROCEDURE_START        PROCEDURE_STOP  PROCEDURE_CODE  \
0  2011-01-02T09:26:36Z  2011-01-02T12:58:36Z     265764009.0   

        PROCEDURE_DESCRIPTION  PROCEDURE_BASE_COST  PROCEDURE_REASON_CODE  \
0  Renal dialysis (procedure)                  903                    NaN   

  PROCEDURE_REASON_DESCRIPTION  
0                          NaN

The Procedures dataframe details medical procedures performed on patients during encounters. Each procedure is linked to a specific patient (PATIENT_ID) and encounter (ENCOUNTER_ID). The dataframe includes the start and stop times of the procedure, the procedure code and description from SNOMED-CT, the base cost of the procedure, and any reasons for the procedure identified by diagnosis codes and descriptions.

<div class="table_component" role="region" tabindex="0">
<table style="border:solid 5px #ca4fec;">
    <caption>
        <p></p>
    </caption>
    <thead>
        <tr>
            <th>Column/Attribute</th>
            <th>Description</th>
            <th>Example</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>
                <ul>
                    <li><strong>PATIENT_ID</strong></li>
                </ul>
            </td>
            <td>Foreign key to the Patient</td>
            <td>
                <div>8fe3348d-477f-541e-96df-bbe5b552104f</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>ENCOUNTER_ID</strong></li>
                </ul>
            </td>
            <td>Foreign key to the Encounter where the procedure was performed</td>
            <td>
                <div>5c504a86-1804-ed37-53ce-32f3d3318ad5</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>PROCEDURE_START</strong></li>
                </ul>
            </td>
            <td>The date and time the procedure was performed (iso8601 UTC Date (yyyy-MM-dd'T'HH:mm'Z'))</td>
            <td>
                <div>2011-01-13T15:41:08Z</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>PROCEDURE_STOP</strong></li>
                </ul>
            </td>
            <td> The date and time the procedure was completed, if applicable (iso8601 UTC Date (yyyy-MM-dd'T'HH:mm'Z'))</td>
            <td>
                <div>2011-01-13T15:56:08Z</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>PROCEDURE_CODE</strong></li>
                </ul>
            </td>
            <td>Procedure code from SNOMED-CT</td>
            <td>
                <div>274804006</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>PROCEDURE_DESCRIPTION</strong></li>
                </ul>
            </td>
            <td>Description of the procedure</td>
            <td>
                <div>Evaluation of uterine fundal height</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>PROCEDURE_BASE_COST</strong></li>
                </ul>
            </td>
            <td>The line-item cost of the procedure</td>
            <td>
                <div>3632</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>PROCEDURE_REASON_CODE</strong></li>
                </ul>
            </td>
            <td>Diagnosis code from SNOMED-CT specifying why this procedure was performed</td>
            <td>
                <div>72892002</div>
            </td>
        </tr>
        <tr>
            <td>
                <ul>
                    <li><strong>PROCEDURE_REASON_DESCRIPTION</strong></li>
                </ul>
            </td>
            <td>Description of the reason code</td>
            <td>
                <div>Normal pregnancy</div>
            </td>
        </tr>
    </tbody>
</table>
</div>

Next, you will receive emails asking you to query the data and answer questions.

---

## Exercise 1: Payers data

![](https://i.ibb.co/zFwqyyX/email.png) **New email received**

<div style="border: dashed 3px red; border-left: solid 6px red; padding: 30px">
Hi,<br/><br/>

I know you're new with us, but I'm sure you'll do great. I need you to pull up **all the information** we have on the 
<span style="background-color:#486ce3;color:white;padding: 0 5px 2px;padding-right:5px;line-height:1em">payers</span> and insurance companies. Thanks!

**Marla**<br/>
Accountant</div>

* <acronym title='In Pandas, to show the full Dataframe, you simply use its name'>HINT</acronym>

Write your answer here:

---

## Exercise 2: Payer contact info

![](https://i.ibb.co/zFwqyyX/email.png) **New email received**

<div style="border: dashed 3px red; border-left: solid 6px red; padding: 30px">
Hello,<br/><br/>

I'm preparing a contact info sheet for the
<span style="background-color:#486ce3;color:white;padding: 0 5px 2px;padding-right:5px;line-height:1em">payers</span> and insurance companies. The only columns I'll need are **names**, **cities**, and **phone numbers**. Thank you in advance.

**Barbara**<br/>
PR and communications specialist</div>

* <acronym title="The syntax for only showing specific columns is Dataframe[['ColumnName1', 'ColumnName2']]">HINT</acronym>

Write your answer here:

---

## Exercise 3: Sample of three patients and top three payers

![](https://i.ibb.co/zFwqyyX/email.png) **New email received**

<div style="border: dashed 3px red; border-left: solid 6px red; padding: 30px">
Good morning,<br/><br/>

I need to get familiar with the <span style="background-color:#85e468;padding: 0 5px 2px;padding-right:5px;line-height:1em">patients</span> and <span style="background-color:#486ce3;color:white;padding: 0 5px 2px;padding-right:5px;line-height:1em">payers</span> data we have. Can you please send me a **random sample of three patients**, as well as **the first three payers** in the payers dataframe?

See you at lunch!

**Jenevive**<br/>
Data analyst</div>

* <acronym title='To generate a random sample of rows, we use the function Dataframe.sample(n) where n is the number of rows. To return the first n rows of a Dataframe, we use the head function: Dataframe.head(n). Likewise, use Dataframe.tail(n) to return the last n rows.'>HINT</acronym>

Write your answer here:

---

## Exercise 4: Columns/Variables in the Patients dataframe

![](https://i.ibb.co/zFwqyyX/email.png) **New email received**

<div style="border: dashed 3px red; border-left: solid 6px red; padding: 30px">
Hi,<br/><br/>

Could you provide a **list of all the column names** in the <span style="background-color:#85e468;padding: 0 5px 2px;padding-right:5px;line-height:1em">patients</span> dataframe, along with a **total count** of the columns? We need this information to evaluate whether we have too many variables in our patient records.

**Samson**<br/>
Privacy officer</div>

* <acronym title="The function Dataframe.info() gives you info on a dataframe's shape (how many rows & columns, just like the Dataframe.shape function), and also a snapshot of the variables, how many null values they have, and their type. Alternatively, if you use the list(Dataframe) function on a Dataframe, it lists the names of the columns in that Dataframe. Then you can simply use the length len(Dataframe) function to count how many columns are in that list.">HINT</acronym>

Write your answer here:

---

## Exercise 5: Lists of latitudes and longitudes

![](https://i.ibb.co/zFwqyyX/email.png) **New email received**

<div style="border: dashed 3px red; border-left: solid 6px red; padding: 30px">
Hello,<br/><br/>

I'm trying to create a map showing where our <span style="background-color:#85e468;padding: 0 5px 2px;padding-right:5px;line-height:1em">patients</span> reside. Can you please send me **two separate series**, one for **latitudes** and one for **longitudes**?

**Dan**<br/>
Market research analyst</div>

* <acronym title="In Pandas, a column can be seen as a one-dimensional array of data, or a Series. To single out a series from a Dataframe, the syntax can either be Dataframe['Series'] (using the column name as an index) or DataFrame.Series (called attribute notation). In this Notebook, we will stick with the first syntax, because it can be used to create new columns, like Dataframe['NewColumn'] = ...,  while the second cannot. Remember this syntax because we will need it to do calculations on any given series.">HINT</acronym>

Write your answer here:

---

## Exercise 6: Unique cities where patients reside

![](https://i.ibb.co/zFwqyyX/email.png) **New email received**

<div style="border: dashed 3px red; border-left: solid 6px red; padding: 30px">
Hi again,<br/><br/>

Thanks for the `LAT` & `LON` lists! Can you please send me a **list of the cities** where our <span style="background-color:#85e468;padding: 0 5px 2px;padding-right:5px;line-height:1em">patients</span> reside **without duplicates**?

**Dan**<br/>
Market research analyst</div>

* <acronym title='Use the Series.unique() function to get an array of its unique non-duplicated values.'>HINT</acronym>

Write your answer here:

---

## Exercise 7: Average, maximum, and minimum base encounter cost

![](https://i.ibb.co/zFwqyyX/email.png) **New email received**

<div style="border: dashed 3px red; border-left: solid 6px red; padding: 30px">
Hi,<br/><br/>

Can you please help me calculate the **average encounter base cost** from the <span style="background-color:#e45151;color:white;padding: 0 5px 2px;padding-right:5px;line-height:1em">encounters</span> data? Also give me the **minimum** and **maximum** base encounter costs on record. Thank you in advance.


**Marla**<br/>
Accountant</div>

* <acronym title='The .mean(), .max(), and .min() functions can be applied directly to Series.'>HINT</acronym>

Write your answer here:

---

## Exercise 8: Single male patients in Norfolk County

![](https://i.ibb.co/zFwqyyX/email.png) **New email received**

<div style="border: dashed 3px red; border-left: solid 6px red; padding: 30px">
Hello,<br/><br/>

Can you please help me figure out how many **single male** <span style="background-color:#85e468;padding: 0 5px 2px;padding-right:5px;line-height:1em">patients</span> we have that live in **Norfolk County**? I'm not looking for a boyfriend or anything, this is for a mental health study about male loneliness :D

**Jenevive**<br/>
Data analyst</div>

* <acronym title="To filter a Dataframe by a condition, we use the syntax Dataframe[Condition]. Here, the Condition is having a Series equal a Value. So the Condition would be Dataframe['Series'] == Value. The full syntax should be Dataframe[Dataframe['Series'] == Value]. Doing this, we get a filtered Dataframe that only shows rows that fullfill the Condition. You can string multiple conditions together using the bitwise operator '&' for 'and', '|' for 'or', and '~' for 'not'.">HINT</acronym>

Write your answer here:

---

## Exercise 9: Marital status count and proportions

![](https://i.ibb.co/zFwqyyX/email.png) **New email received**

<div style="border: dashed 3px red; border-left: solid 6px red; padding: 30px">
Hi again,<br/><br/>

Speaking of marital status, can you give me a **count** of how many single vs married <span style="background-color:#85e468;padding: 0 5px 2px;padding-right:5px;line-height:1em">patients</span> we have? I'll also need their **proportions** or percentages. Don't forget to take into consideration patients who are missing this info!

**Jenevive**<br/>
Data analyst</div>

* <acronym title="To count the number of occurrences in a series, use the .value_counts() function. The function has two useful parameters: .value_counts(dropna=False) will count even the null values, and .value_counts(normalize=True) will give proportions instead of counts.">HINT</acronym>

Write your answer here:

---

## Exercise 10: Missing entry

![](https://i.ibb.co/zFwqyyX/email.png) **New email received**

<div style="border: dashed 3px red; border-left: solid 6px red; padding: 30px">
Hi again,<br/><br/>

Thank you for the data on single patients! I noticed from your last email that one of the <span style="background-color:#85e468;padding: 0 5px 2px;padding-right:5px;line-height:1em">patients</span> didn't have any marital info. Can you please send me the `PATIENT_ID` of the patient lacking the marital status so we can follow up with them?

**Jenevive**<br/>
Data analyst</div>

* <acronym title="The function .isna() or .isnull() (isnull is just an alias of isna) can be used to check if a value is Null or NaN. If it is passed as a condition: Dataframe[Dataframe['Series'].isna()], it will return a dataframe with all the rows that have a Null value for that column.">HINT</acronym>

Write your answer here:

---

## Exercise 11: Race and ethnicity pivot table

![](https://i.ibb.co/zFwqyyX/email.png) **New email received**

<div style="border: dashed 3px red; border-left: solid 6px red; padding: 30px">
Hi,<br/><br/>

Do you happen to know **how many** <span style="background-color:#85e468;padding: 0 5px 2px;padding-right:5px;line-height:1em">patients</span> we have served **for each combination of the `RACE` and `ETHNICITY`** variables? If you can display it in a **pivot table**, that would be great. This info is useful because we need to ensure equitable access to health care across the board.

**Zoey**<br/>
CEO</div>

The expected result looks like this:

<pre>
RACE	             asian	black	hawaiian    native	other	white
ETHNICITY
    
hispanic	       14	   49	    2	       3	   4	  119
nonhispanic	       77	  114	   11	       8	  12	  561
</pre>

* <acronym title="The syntax for a pivot table is Dataframe.pivot_table(Dataframe, index='Series1', columns='Series2', aggfunc='size'). The aggfunc parameter can be any function we want, but here we just need the frequency or count. Alternatively, since both RACE and ETHNICITY are categorical variables, we can use the crosstab function like this: pd.crosstab(Dataframe['Series1'], Dataframe['Series2']). Crosstab can only do frequencies, and no other functions, unlike pivot_table. Also, note that it's a function of Pandas, hence the pd.crosstab(), and not a dataframe method like .pivot_table().">HINT</acronym>

Write your answer here:

---

## Exercise 12: Birthday of patients

![](https://i.ibb.co/zFwqyyX/email.png) **New email received**

<div style="border: dashed 3px red; border-left: solid 6px red; padding: 30px">
Good afternoon,<br/><br/>

Do we have any <span style="background-color:#85e468;padding: 0 5px 2px;padding-right:5px;line-height:1em">patients</span> who are celebrating their **birthday on March 15**? If so, can you please send me a sheet with their `PATIENT_ID`, **full name** (new column), and `BIRTH_DATE`.

Thanks!

**Martha**<br/>
Patient outreach officer</div>

The correct result should look like this:
<pre>
	PATIENT_ID	                                FULL_NAME	         BIRTH_DATE
231	381e16cd-86c8-d369-cd76-2dd9e0b5faf5	Fredrick Gutmann	 1939-03-15
593	358775ff-7a45-d881-7a3d-e26b06732f1a	Wm Morar	         1964-03-15
626	f39ec5de-b029-c6d4-2738-696ecab81a79	Will Hintz	         1990-03-15
969	d684571e-a784-ef61-429e-06fa0d2b1637	Chris Fisher             1924-03-15
</pre>

* <acronym title="Before working with a Series that contains dates and times, it's important to ensure that the Series is correctly recognized as a datetime type. This can be done using the pandas function .to_datetime() like this: Series = pd.to_datetime(Series). Then, we can perform date-related operations like extracting components (e.g., year, month, day) like this: Series.dt.day, Series.dt.month, or Series.dt.year. Lastly, we can concatenate two Series into one using the + operator (if they are both strings), but don't forget to add a space in between.">HINT</acronym>

Write your answer here:

---

## Exercise 13: Patients with the initials A.C.

![](https://i.ibb.co/zFwqyyX/email.png) **New email received**

<div style="border: dashed 3px red; border-left: solid 6px red; padding: 30px">
Hi,<br/><br/>

I'm looking for a patient with the **initials A. C.**, but I don't know anything else about them. Can you retrieve all the <span style="background-color:#85e468;padding: 0 5px 2px;padding-right:5px;line-height:1em">patients</span> that have these initials?

**Jenevive**<br/>
Data analyst</div>

The correct answer should show 4 patients: Argentina Corwin, Anthony Cronin, Armando Carrasquillo, and Arturo Cuellar.

* <acronym title="Like datetime series that have the accessor .dt, string series have the accessor .str. It comes with many useful functions like .str.startswith().">HINT</acronym>

Write your answer here:

---

## Exercise 14: Patients from Germany

![](https://i.ibb.co/zFwqyyX/email.png) **New email received**

<div style="border: dashed 3px red; border-left: solid 6px red; padding: 30px">
Good morning,<br/><br/>

I'm looking for <span style="background-color:#85e468;padding: 0 5px 2px;padding-right:5px;line-height:1em">patients</span> that were born in Germany. The `BIRTHPLACE` column ends with an alpha-2 code that denotes the country of birth. The code for Germany is **DE**. I hope this is helpful.

**Dan**<br/>
Market research analyst</div>

* <acronym title="Just like .str.startswith() that we used in the last exercise, we can use .str.endswith().">HINT</acronym>

The correct answer should show three patients: Ike Becker, Lawerence Vandervort, and Terresa Nader.

Write your answer here:

---

## Exercise 15: Countries of birth

![](https://i.ibb.co/zFwqyyX/email.png) **New email received**

<div style="border: dashed 3px red; border-left: solid 6px red; padding: 30px">
Hi again,<br/><br/>

Thank you for the info on the German patients. Can you provide a breakdown of the **8 countries** where most of our <span style="background-color:#85e468;padding: 0 5px 2px;padding-right:5px;line-height:1em">patients</span> were born?

**Dan**<br/>
Market research analyst</div>

The correct result should look like this:

<pre>
    BIRTH_COUNTRY
US    845
DM     17
MX     16
PR     13
CO      9
CN      8
VN      7
HT      5
</pre>

* <acronym title="In this exercise, we'll need to do some slicing and dicing. If we look at the BIRTHPLACE column, the country code always comes last. So we can split the string at the spaces, and take the last part. We can do that using .str.split() and .str[index], knowing that the last index in Python can be referred to by -1. After putting the country codes in their own new column, we can then use .groupby().size() to group the data using that column.">HINT</acronym>

Write your answer here:

---

## Exercise 16: Depression-related procedures

![](https://i.ibb.co/zFwqyyX/email.png) **New email received**

<div style="border: dashed 3px red; border-left: solid 6px red; padding: 30px">
Hi,<br/><br/>

Do you happen to know how many <span style="background-color:#ca4fec;color:white;padding: 0 5px 2px;padding-right:5px;line-height:1em">procedures</span> we have that are **related to depression**? Which one is the **most used**? 

The PHQ-2 (two-item questionnaire) is used to screen for depression, but after that, the patient needs to undergo the PHQ-9 (nice-item questionnaire) to gauge the severity of the depression. I'm afraid not many patients undergo PHQ-9 after their first initial screening. Can you please confirm this?

**Barbara**<br/>
Mental health specialist</div>

The correct answer should look like this:

<pre>
    PROCEDURE_DESCRIPTION
Depression screening                                                                   3804
Depression screening using Patient Health Questionnaire Two-Item score (procedure)     3614
Depression screening using Patient Health Questionnaire Nine Item score (procedure)     206
</pre>

* <acronym title="Another useful function we can use with the .str accessor is .contains(). We can use it to return rows where the condition of a string containing a sub-string is satisfied. By default, the function is case sensitive, unless we add the argument case=Fase. Attention: this function will not work if the series contains a null value.">HINT</acronym>

Write your answer here:

---

## Exercise 17: Full addresses of payers

![](https://i.ibb.co/zFwqyyX/email.png) **New email received**

<div style="border: dashed 3px red; border-left: solid 6px red; padding: 30px">
Good morning,<br/><br/>

Can you help me prepare the **full addresses** of <span style="background-color:#486ce3;color:white;padding: 0 5px 2px;padding-right:5px;line-height:1em">payers</span> and health insurance companies? I need it for our printed newsletter. The format should be **Street address, City, State, Zip code**.

**Barbara**<br/>
PR and communications specialist</div>

The correct answer should look like this:

<pre>
        PAYER_NAME              FULL_ADDRESS
0	Dual Eligible           7500 Security Blvd, Baltimore, MD, 21244
1	Medicare                7500 Security Blvd, Baltimore, MD, 21244
2	Medicaid                7500 Security Blvd, Baltimore, MD, 21244
3	Humana                  500 West Main St, Louisville, KY, 40018
4	Blue Cross Blue Shield	Michigan Plaza, Chicago, IL, 60007
5	UnitedHealthcare        9800 Healthcare Lane, Minnetonka, MN, 55436
6	Aetna                   151 Farmington Ave, Hartford, CT, 6156
7	Cigna Health            900 Cottage Grove Rd, Bloomfield, CT, 6002
8	Anthem                  220 Virginia Ave, Indianapolis, IN, 46204
9	NO_INSURANCE            NaN
</pre>

* <acronym title="There are many ways to concatenate a bunch of columns into a new one. We've done it before, but this time we have string and flot columns, so we need to convert them to strings before joining them together using +. That can be done with str(Series) or Series.astype(str).">HINT</acronym>

Write your answer here:

---

## Exercise 18: Group by encounter class

![](https://i.ibb.co/zFwqyyX/email.png) **New email received**

<div style="border: dashed 3px red; border-left: solid 6px red; padding: 30px">
Hello,<br/><br/>

Can you **group** the <span style="background-color:#e45151;color:white;padding: 0 5px 2px;padding-right:5px;line-height:1em">encounters</span> by **`ENCOUNTER_CLASS`** and **count** how many encounters fall into each class? Also, in a separate dataframe, please **calculate the sum of the `TOTAL_CLAIM_COST`** for each class. Please make sure to sort the results in descending order (from big to small).

Thanks!

**Leo**<br/>
Statistician</div>

The correct answer should look like this:

<pre>
ENCOUNTER_CLASS
ambulatory    12537
outpatient     6300
urgentcare     3666
emergency      2322
wellness       1931
inpatient      1135

ENCOUNTER_CLASS
ambulatory    36283440.84
urgentcare    23349349.96
outpatient    14094967.12
emergency     10750051.53
inpatient      8809134.34
wellness       8227431.73
</pre>

* <acronym title="As we've seen before, you can use .groupby().size() for frequency or count. To get the sums, we can use the .sum() function like this Dataframe.groupby(Series1)[Series2].sum(), Series2 being the series we want to apply the sum function to.">HINT</acronym>

Write your answer here:

---

## Exercise 19: Group by encounter year

![](https://i.ibb.co/zFwqyyX/email.png) **New email received**

<div style="border: dashed 3px red; border-left: solid 6px red; padding: 30px">
Hi,<br/><br/>

Wonderful, thank you! Can you please do the **same grouping** but using the **year** of the <span style="background-color:#e45151;color:white;padding: 0 5px 2px;padding-right:5px;line-height:1em">encounters</span> instead of the class? I'll need the **count** as well as the **sum** of the `TOTAL_CLAIM_COST` for each year. In both instances, sort by year chronologically.

**Leo**<br/>
Statistician</div>

The correct answer should look like this:

<pre>
YEAR
2011    1336
2012    2106
2013    2495
2014    3885
2015    2469
2016    2451
2017    2360
2018    2292
2019    2228
2020    2519
2021    3530
2022     220

YEAR
2011     4405936.73
2012     9058314.10
2013    10105272.87
2014    12011662.17
2015    10124652.89
2016     8386701.12
2017     9596894.47
2018     8746818.37
2019     8915338.78
2020    10294077.96
2021     9177232.00
2022      691474.06  
</pre>

* <acronym title="First we need to convert to datetime type using pd.to_datetime(), then use the accessor .dt to extract the years and put them in their own column before applying .groupby().">HINT</acronym>

Write your answer here:

---

## Exercise 19: Group by encounter year (cont.)

![](https://i.ibb.co/zFwqyyX/email.png) **New email received**

<div style="border: dashed 3px red; border-left: solid 6px red; padding: 30px">
Hi,<br/><br/>

Thanks! I hate to ask, but can you please put the **count** and the **sum** calculations from the last email in their own columns in the same dataframe?

**Leo**<br/>
Statistician</div>

The correct answer should look like this:

<pre>
        Count           Sum
YEAR		
2011	1336            4405936.73
2012	2106            9058314.10
2013	2495            10105272.87
2014	3885            12011662.17
2015	2469            10124652.89
2016	2451            8386701.12
2017	2360            9596894.47
2018	2292            8746818.37
2019	2228            8915338.78
2020	2519            10294077.96
2021	3530            9177232.00
2022	220             691474.06
</pre>

* <acronym title="The way we used .groupby() before, it returns a series, not a dataframe. We can use the .agg() function we've seen before to generate two columns, one with the frequency or count, and one with the sums. The syntax would be Dataframe.groupby(Series).agg(NewColumnName1=('Series1':'count'), NewColumnName2=('Series2':'sum')). Try to apply the count function on a column that has no null values like the ID column.">HINT</acronym>

Write your answer here:

---

## Exercise 20: Group by encounters year and month

![](https://i.ibb.co/zFwqyyX/email.png) **New email received**

<div style="border: dashed 3px red; border-left: solid 6px red; padding: 30px">
Hi,<br/><br/>

This is the last favor I'll ask and I'll get out of your hair (for today at least). I need to figure out the months that get very busy. Can you please **group** the <span style="background-color:#e45151;color:white;padding: 0 5px 2px;padding-right:5px;line-height:1em">encounters</span> **count by year** and then **month**?

**Leo**<br/>
Statistician</div>

The correct result should look like this:

<pre>
    	YEAR	MONTH	Encounters Count
0	2011	1	86
1	2011	2	86
2	2011	3	129
3	2011	4	113
4	2011	5	131
...	...	...	...
129	2021	10	225
130	2021	11	267
131	2021	12	231
132	2022	1	219
133	2022	2	1
    
</pre>

* <acronym title="We can group by multiple columns like this .groupby(['Series1', 'Series2']). After that, if we want to return a dataframe instead if a series, we can use the .reset_index() function, and even give a name to the column that shows the frequency or count by passing the argument name='ColumnName'.">HINT</acronym>

Write your answer here:

---

## Exercise 21: COVID-19 cases

![](https://i.ibb.co/zFwqyyX/email.png) **New email received**

<div style="border: dashed 3px red; border-left: solid 6px red; padding: 30px">
Hi,<br/><br/>

We are conducting a study about the effects of the COVID-19 pandemic. Can you please give me a breakdown of all the <span style="background-color:#e45151;color:white;padding: 0 5px 2px;padding-right:5px;line-height:1em">encounters</span> related to COVID-19, **grouped by the description of the encounter**? I'm sure the `ENCOUNTER_REASON_DESCRIPTION` mentions 'COVID' whenever the case is COVID-19 related.

**Jason**<br/>
Director of clinical research</div>

The correct answer should look like this:

<pre>
    	ENCOUNTER_DESCRIPTION	                             COVID-19 Cases
0	Admission to intensive care unit (procedure)            7
1	Death Certification	                                2
2	Hospital admission for isolation (procedure)	       17
</pre>

* <acronym title="Nothing new here, we'll be using the functions .isna(), .str.contains(), and .groupby().size().reset_index() like before.">HINT</acronym>

Write your answer here:

---

## Exercise 22: COVID-19 ICU duration

![](https://i.ibb.co/zFwqyyX/email.png) **New email received**

<div style="border: dashed 3px red; border-left: solid 6px red; padding: 30px">
Hi again,<br/><br/>

Thank you for that information. **For how long** did we care for the seven COVID-19 cases that were **admitted to the Intensive Care Unit** (ICU)?

Please send me a dataframe with this structure, populated with data taken from the <span style="background-color:#e45151;color:white;padding: 0 5px 2px;padding-right:5px;line-height:1em">encounters</span> dataframe:
<style>
.table_component {
    overflow: auto;
    width: 100%;
}

.table_component table {
    border: 1px solid #000000;
    height: 100%;
    width: 100%;
    table-layout: auto;
    border-collapse: collapse;
    border-spacing: 1px;
    text-align: left;
}

.table_component caption {
    caption-side: top;
    text-align: left;
}

.table_component th {
    border: 1px solid #dededf;
    background-color: #eceff1;
    color: #000000;
    padding: 5px;
}

.table_component td {
    border: 1px solid #dededf;
    background-color: #ffffff;
    color: #000000;
    padding: 5px;
}
</style>
<div class="table_component" role="region" tabindex="0">
<table>
     <thead>
        <tr>
            <th>PATIENT_ID</th>
            <th>
                ENCOUNTER_DESCRIPTION
            </th>
            <th>ENCOUNTER_REASON_DESCRIPTION</th>
            <th>DURATION_IN_DAYS</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>
                <div><b><br></b></div>
            </td>
            <td></td>
            <td><br></td>
            <td></td>
        </tr>
        <tr>
            <td></td>
            <td></td>
            <td></td>
            <td></td>
        </tr>
    </tbody>
</table>

**Jason**<br/>
Director of clinical research</div>

The correct answer should look like this:

<pre>
ENCOUNTER_START	                ENCOUNTER_STOP	          ENCOUNTER_DESCRIPTION ENCOUNTER_REASON_DESCRIPTION    DURATION
23083	2020-08-20 01:55:34+00:00 2020-08-29 08:25:58+00:00	Admission to intensive ...	COVID-19	9 days 06:30:24
23441	2020-10-09 07:57:23+00:00 2020-10-15 10:41:30+00:00	Admission to intensive ...	COVID-19	6 days 02:44:07
23710	2020-11-10 10:02:48+00:00 2020-11-15 18:07:47+00:00	Admission to intensive ...	COVID-19	5 days 08:04:59
23898	2020-12-03 07:26:52+00:00 2020-12-08 04:21:35+00:00	Admission to intensive ...	COVID-19	4 days 20:54:43
24132	2020-12-30 23:06:36+00:00 2021-01-03 04:45:25+00:00	Admission to intensive ...	COVID-19	3 days 05:38:49
24251	2021-01-15 23:26:52+00:00 2021-01-22 22:42:01+00:00	Admission to intensive ...	COVID-19	6 days 23:15:09
24285	2021-01-19 21:18:02+00:00 2021-01-28 05:46:41+00:00	Admission to intensive ...	COVID-19	8 days 08:28:39
</pre>

* <acronym title="After converting the date and time columns to datetype type using pd.to_datetime(), we can then create a new column that contains the difference between the start and the stop like this: Series1 = Series2 - Series3">HINT</acronym>

Write your answer here:

---

## Exercise 23: The five oldest living patients

![](https://i.ibb.co/zFwqyyX/email.png) **New email received**

<div style="border: dashed 3px red; border-left: solid 6px red; padding: 30px">
Hi,<br/><br/>

I'm working on a survey for our senior <span style="background-color:#85e468;padding: 0 5px 2px;padding-right:5px;line-height:1em">patients</span>. Can you please pull the data of the **five oldest (still living) patients** we have ever served? I'll need `PATIENT_ID`, `FIRST_NAME`, `LAST_NAME`, `BIRTH_DATE`, and Please **add an** `AGE` **column** to the result that displays the age in years.

Thanks!

**Martha**<br/>
Patient outreach officer</div>

The correct result should look like this (the AGE column values will be different based on the day you run this cell)

<pre>

    	PATIENT_ID	                        FIRST_NAME  LAST_NAME	BIRTH_DATE         AGE
868	fc81fad8-86e1-4df2-0613-25019adf3e5d	Elwanda	    Casper	1922-03-24	   102.335387
547	de1f6a21-bdd3-0ccd-8a3c-b06e98a13cf2	Lulu	    Carter	1922-03-30	   102.318960
859	d1629b63-3486-1537-a088-cd641d231e86	Lazaro	    Osinski	1922-04-13	   102.280630
876	b2b5f8d2-9ba8-f517-5ce5-3f246a0eee52	Horace	    Cremin	1922-04-15	   102.275154
553	c6b2ae50-d8d6-014f-bfff-27afbeb2e1e8	Trent	    Graham	1922-06-18	   102.099932
</pre>

* <acronym title="To get today's date, you can use pd.to_datetime('today'). After the usual time subtraction using -, convert the result to days using .dt.days, then you can divide it by 365 to get the years, or 365.25 if you want to take into account leap years. ">HINT</acronym>

Write you answer here:

---

## Exercise 24: Age groups for patients

![](https://i.ibb.co/zFwqyyX/email.png) **New email received**

<div style="border: dashed 3px red; border-left: solid 6px red; padding: 30px">
Hi again,<br/><br/>

Thank you for the senior patients list! 

I have another favor to ask. Can you please help me put the <span style="background-color:#85e468;padding: 0 5px 2px;padding-right:5px;line-height:1em">patients</span> into **age groups** like this: 0-18, 19-35, 36-50, 51-70, and 71+? Give me a **count** of our patients under each group, as well as their **proportions** (or percentages). We tend to have more senior patients than young ones, and I want to confirm that with some data.

**Martha**<br/>
Patient outreach officer</div>

The correct result should look like this (the values might change based on the date this cell is executed):

<pre>
AGE_GROUP
0-18       0
19-35     25
36-50    164
51-70    234
71+      505

AGE_GROUP
0-18      0.000000
19-35     2.566735
36-50    16.837782
51-70    24.024641
71+      51.848049
</pre>

* <acronym title="To cut a series into bins, we use the Pandas function pd.cut(). The syntax is: Series1 = pd.cut(Series2, bins=[..], labels=[..]). The bins parameter is a list of values that describe where the cut should happen, and the label is a list of labels for each cut. Hint 1: we need to have 6 values in bins, and only 5 in labels. Hint 2: the last value in bins needs to take into account the biggest age, so it preferably should be a variable or function, not a hard value. Finally, while grouping by .size() is straightforward, you'll need a custom lambda function to calculate the proportions, which you'll be applying with the .apply() function.">HINT</acronym>

Write your answers here:

---

## Exercise 25: Patient IDs in Patients & Encounters

![](https://i.ibb.co/zFwqyyX/email.png) **New email received**

<div style="border: dashed 3px red; border-left: solid 6px red; padding: 30px">
Good morning,<br/><br/>

I need you to help me verify if all the patients we have on record have had an encounter at some point. In other words, check if all the **unique `PATIENT_ID`** values in <span style="background-color:#85e468;padding: 0 5px 2px;padding-right:5px;line-height:1em">patients</span> **are present** in <span style="background-color:#e45151;color:white;padding: 0 5px 2px;padding-right:5px;line-height:1em">encounters</span>, and **vice-versa**.

**John**<br/>
Data integrity officer</div>

* <acronym title="You can pass a Series into a set() function like this: set(Series). This turns it into a set, which is a collection data type with only unique values. You can either compare two sets: set1 == set2 to make sure they hold the same unique values, or you can use set1.issubset(set2) to make sure all the values in set1 are present in set2, and the other way around, set2.issubset(set1). If both are True, then the sets are identical. ">HINT</acronym>

Write your answer here:

---

## Exercise 26: Patients with the initials A. C. (cont.)

![](https://i.ibb.co/zFwqyyX/email.png) **New email received**

<div style="border: dashed 3px red; border-left: solid 6px red; padding: 30px">
Hi,<br/><br/>

Remember when you sent me the four <span style="background-color:#85e468;padding: 0 5px 2px;padding-right:5px;line-height:1em">patients</span> with the initials A. C.; Argentina Corwin, Anthony Cronin, Armando Carrasquillo, and Arturo Cuellar? Now I have an additional clue about the patient I am looking for to narrow the search down: they only have **one single encounter** on record on the <span style="background-color:#e45151;color:white;padding: 0 5px 2px;padding-right:5px;line-height:1em">encounters</span> dataframe. Can you help me find them? Thanks!

**Jenevive**<br/>
Data analyst</div>

The correct answer should look like this:

<pre>
FIRST_NAME
Anthony       1
Argentina    28
Armando      28
Arturo        9
</pre>

* <acronym title="To merge two dataframes on a commun column, we use the Pandas function pd.merge(). The syntax is NewDataframe = pd.merge(Dataframe1, Dataframe2, on='CommonColumn'). Additionally, we can pass a how parameter that dictates what happens to the non-common rows on both sides, but since the last exercise showed the two dataframes patients and encounters sharing the exact same IDs, we don't need to worry about that now.">HINT</acronym>

Write your answer here:

---

## Exercise 27: Best health insurance

![](https://i.ibb.co/zFwqyyX/email.png) **New email received**

<div style="border: dashed 3px red; border-left: solid 6px red; padding: 30px">
Greetings,<br/><br/>

Could you assist me in creating a ranking of <span style="background-color:#486ce3;color:white;padding: 0 5px 2px;padding-right:5px;line-height:1em">payers</span> and health insurance companies based on the quality of their coverage? The ranking should be based on two key metrics related to how well they pay for patient <span style="background-color:#e45151;color:white;padding: 0 5px 2px;padding-right:5px;line-height:1em">encounters</span>:

* The average percentage of costs they covered.
* The number of times they refused to cover any portion of an encounter (0% coverage).


**Marla**<br/>
Accountant</div>

The correct answer should look like this:

<pre>
    
        PAYER_NAME              AVERAGE_PERCENTAGE  NUMBER_OF_REFUSALS
6	Medicaid                74.550858           0
7	Medicare                62.950728           0
4	Dual Eligible           45.901624           123
2	Blue Cross Blue Shield  32.021687           336
9	UnitedHealthcare        2.613364            817
5	Humana                  1.255674            1036
0	Aetna                   0.502589            898
3	Cigna Health            0.201057            789
8	NO_INSURANCE            0.000000            8795
1	Anthem                  0.000000            703
</pre>

* <acronym title="After merging the payers and encounters dataframes using pd.merge(), you'll need to create a column 'AVERAGE_PERCENTAGE' that's basically the payer coverage divided by the total claim cost, multiplied by 100. There are many ways to do this, but I suggest you try the function Dataframe.eval(). Like .query(), it takes a string as input, and you can use the column names as variables. Lastly, we need to sort the values by both the average percentage AND the number of refusals, to get a ranking of the best payers.">HINT</acronym>

Write your answer here:

---

## Exercise 28: The six most expensive procedures

![](https://i.ibb.co/zFwqyyX/email.png) **New email received**

<div style="border: dashed 3px red; border-left: solid 6px red; padding: 30px">
Hi,<br/><br/>

Can you please help me figure out the **6 most expensive** <span style="background-color:#ca4fec;color:white;padding: 0 5px 2px;padding-right:5px;line-height:1em">procedures</span>
 our hospital is providing? Give me the **procedure description**, the **base cost**, and the **reason** the procedure was necessary. The same procedure can have different base costs depending on multiple factors, so give me the most expensive base cost of each of the 6 most expensive procedure.

**Zoey**<br/>
CEO</div>

The correct answer should look like this:

<pre>
    	PROCEDURE_DESCRIPTION	                    PROCEDURE_BASE_COST	 PROCEDURE_REASON_DESCRIPTION
41089	Admit to ICU (procedure)	                  289531	 Septic shock (disorder)
37398	Coronary artery bypass grafting	                  67168	         Myocardial Infarction
39780	Electrical cardioversion	                  44532	         Atrial Fibrillation
26032	Partial resection of colon	                  38749	         Malignant tumor of colon
6499	Insertion of biventricular implantable cardiov... 38711	         Cardiac Arrest
41268	Hemodialysis (procedure)	                  37213	         NaN
</pre>

* <acronym title="We want to group by the procedure descriptions and then select the row with the highest base cost for each procedure. To achieve this, you'll use the .idxmax() function. When appended after a .groupby() operation, where you might normally use .size() or .sum(), .idxmax() will return the index labels of the rows with the maximum value in each group. You can then use this output with the .loc[] function to retrieve those rows: DataFrame.loc[idxmax_output, ['column1', 'column2', ...]]. Similarly, you can use .idxmin() to select the rows with the lowest values.">HINT</acronym>

Write your answer here:

---

## Exercise 29: Coronary Artery Bypass Grafting health coverage

![](https://i.ibb.co/zFwqyyX/email.png) **New email received**

<div style="border: dashed 3px red; border-left: solid 6px red; padding: 30px">
Hello,<br/><br/>

I was tasked by Zoey, our CEO, to investigate the health coverage/insurance situation of the patients who underwent the **Coronary Artery Bypass Grafting (CABG) procedure**, one of the most expensive procedures we provide. We are working with our non-profit partners that promote cardiovascular health to help patients with no health insurance cover their costs.

Can you please provide me with the **names** of the <span style="background-color:#85e468;padding: 0 5px 2px;padding-right:5px;line-height:1em">patients</span>, the <span style="background-color:#ca4fec;color:white;padding: 0 5px 2px;padding-right:5px;line-height:1em">procedure</span> **description**, the **total claim cost**, the <span style="background-color:#486ce3;color:white;padding: 0 5px 2px;padding-right:5px;line-height:1em">payers</span>
 **name**, and the **payer coverage** for all <span style="background-color:#e45151;color:white;padding: 0 5px 2px;padding-right:5px;line-height:1em">encounters</span> related to CABG? 
 
 ps. The `PROCEDURE_CODE` for CABG is 232717009.

**Martha**<br/>
Patient outreach officer</div>

The correct answer should look like this:

<pre>
	   FIRST_NAME LAST_NAME   PROCEDURE_DESCRIPTION	        TOTAL_CLAIM_COST  PAYER_NAME	PAYER_COVERAGE
18636	Josiah	 Purdy        Coronary artery bypass grafting	62625.44	Medicare	49961.53
18706	Jay	 Collins      Coronary artery bypass grafting	100055.91	Medicare	79953.92
26242	Mariano	 O'Kon	      Coronary artery bypass grafting	1244.21	        NO_INSURANCE	0.00
30620	Ignacio	 Hermiston    Coronary artery bypass grafting	78693.97	Anthem	        0.00
31543	Lorenzo	 Prieto	      Coronary artery bypass grafting	59785.85	NO_INSURANCE	0.00
31892	Ignacio	 Hermiston    Coronary artery bypass grafting	47448.53	NO_INSURANCE	0.00
35419	Garfield Block        Coronary artery bypass grafting	43978.19	Medicare	35086.55
36440	Junior	 Padberg      Coronary artery bypass grafting	52689.25	Medicaid	49883.79
37398	Elvis	 Hackett      Coronary artery bypass grafting	85905.65	Medicare	68642.02
</pre>

* <acronym title="This is just an exercise on merging 4 dataframes. You already know everything you need to solve it. You can merge the dataframes two by two to make it easier.">HINT</acronym>

Write your answer here:

---

## Exercise 30: Deaths after complicated procedures

![](https://i.ibb.co/zFwqyyX/email.png) **New email received**

<div style="border: dashed 3px red; border-left: solid 6px red; padding: 30px">
Hi,<br/><br/>

Can you please help me research the <span style="background-color:#e45151;color:white;padding: 0 5px 2px;padding-right:5px;line-height:1em">encounters</span> where <span style="background-color:#85e468;padding: 0 5px 2px;padding-right:5px;line-height:1em">patients</span> died shortly before, during or after complicated encounters?

In our standards, we adopt a **30-day mortality metric** when it comes to heart problems; meaning if a patient dies within 30 days before or after a heart-related encounter, the mortality is attributed to complications from that encounter. For cancers, the metric is **90 days**. For all other medical encounters, you can safely use a **10-day metric**.

To help you out, these are keywords related to heart disease and cancers that you may use to filter the **encounters descriptions** and the **encounters reason descriptions**: 

<br/>**Heart disease**: `heart failure`, `cardiac arrest`, `stroke`, `coronary`, `atrial fibrillation`, `myocardial infarction`, `thrombectomy`
<br/>**Cancer**: `cancer`, `carcinoma`, `malignant tumor`, `neoplasm`, `chemotherapy`

Remember that each patient has multiple encounters on file, so you want to concentrate on the last encounter closer to their `DEATH_DATE`.

Please send me three separate dataframes like this, one for each of the three death causes (Hearth, Cancer, and Else):

<div class="table_component" role="region" tabindex="0">
<table>
    <caption>
    </caption>
    <thead>
        <tr>
            <th>FIRST_NAME</th>
            <th>
                LAST_NAME
            </th>
            <th>ENCOUNTER_STOP</th>
            <th>DEATH_DATE</th>
            <th>DEATH_MARGIN</th>
            <th>ENCOUNTER_DESCRIPTION</th>
            <th>ENCOUNTER_REASON_DESCRIPTION</th>
            <th>CAUSE</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>
                <div>...</div>
            </td>
            <td>...</td>
            <td>...</td>
            <td>...</td>
            <td>...</td>
            <td>...</td>
            <td>...</td>
            <td>...</td>
        </tr>
    </tbody>
</table>
</div>

`DEATH_MARGIN` is the number of days between `ENCOUNTER_STOP` and `DEATH_DATE` (could be positive or negative, so use the absolute value when filtering).

Thanks!

**Jason**<br/>
Director of clinical research</div>

The correct answer should look like this:

* For Heart disease, the first 3 columns are of patients Wilmer Blanda (8 days before), Gaston Schneider (6 days after), and Jay Collins (8 days before).
* For Cancers, the first 3 columns are of patients Olin Kerluke, Ike Becker, and Kurtis Bartell, all died 7 days before the end of the last encounter.
* For other causes, the first 3 columns are of patients Corrin Heaney (9 days after), Cory Smitham (1 day before), and Leana Boyer (5 days after).	


* <acronym title="To solve this exercise, we want to search for keywords in two string columns: ENCOUNTER_DESCRIPTION and ENCOUNTER_REASON_DESCRIPTION. We could use the .str accessor with the .contains() function, but there is an easier way: set up two lists of keywords, one for heart and one for cancer. Define a function that combines the two descriptions into one combined string, and then uses a 'for' loop that returns 'Heart' when any term in the heart list is in the combined description, likewise for 'Cancer', and finally returns 'Else' if neither of these conditions is true. Apply that function to a new column like this Dataframe['NewColumn'] = Dataframe.apply(MyFunction, axis=1). ATTENTION: ENCOUNTER_REASON_DESCRIPTION has null values sometimes, so this needs to be addressed before proceeding with the previous operation. This exercise also requires the use of datetime methods, .groupby(), .idxmax(), .loc, and .abs(). Bonus points if you use .query() as well.">HINT</acronym>

Write your answer here:

* Heart disease deaths:

---

## Conclusion:

I hope you enjoyed these exercises! If you learned something new, please leave a comment and share your thoughts on my [Medium article here](). I'd love to hear any creative solutions you came up with - if you have a unique approach, I might feature it in the article with a shoutout to you. 

If you're looking for more advanced challenges, I might create some in a future article. Next, I may work on the SQL equivalent of these exercises, or data visualization exercises. Let me know in the comments what you want to see next. 

Stay tuned for more!